sw/device/lib/crypto/drivers/keymgr.c

```diff
@@ -0,0 +1,257 @@
+  HARDENED_TRY(keymgr_is_idle());
+
+  // Set the control register to generate a software-visible key.
+  WRITE_CTRL(NONE, GENERATE_SW);
+
+  // Check that the version does not exceed the max owner key version.
+  uint32_t max_version =
+      abs_mmio_read32(kBaseAddr + KEYMGR_MAX_OWNER_KEY_VER_SHADOWED_REG_OFFSET);
+  if (diversification.version > max_version) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Start the operation and wait for it to complete.
+  keymgr_start(diversification);
+  HARDENED_TRY(keymgr_wait_until_done());
+
+  // Collect output.
+  // TODO: for SCA hardening, randomize the order of these reads.
+  for (size_t i = 0; i < kKeymgrOutputShareNumWords; i++) {
+    key->share0[i] =
+        abs_mmio_read32(kBaseAddr + KEYMGR_SW_SHARE0_OUTPUT_0_REG_OFFSET +
+                        (i * sizeof(uint32_t)));
+  }
+  for (size_t i = 0; i < kKeymgrOutputShareNumWords; i++) {
+    key->share1[i] =
+        abs_mmio_read32(kBaseAddr + KEYMGR_SW_SHARE1_OUTPUT_0_REG_OFFSET +
+                        (i * sizeof(uint32_t)));
+  }
+
+  return OTCRYPTO_OK;
+}
+
+status_t keymgr_generate_key_aes(keymgr_diversification_t diversification) {
+  HARDENED_TRY(keymgr_is_idle());
+
+  // Set the control register to generate an AES key.
+  WRITE_CTRL(AES, GENERATE_HW);
+
+  // Start the operation and wait for it to complete.
+  keymgr_start(diversification);
+  return keymgr_wait_until_done();
+}
+
+status_t keymgr_generate_key_kmac(keymgr_diversification_t diversification) {
+  HARDENED_TRY(keymgr_is_idle());
+
+  // Set the control register to generate a KMAC key.
+  WRITE_CTRL(KMAC, GENERATE_HW);
+
+  // Start the operation and wait for it to complete.
+  keymgr_start(diversification);
+  return keymgr_wait_until_done();
+}
+
+status_t keymgr_generate_key_otbn(keymgr_diversification_t diversification) {
+  HARDENED_TRY(keymgr_is_idle());
+
+  // Set the control register to generate an OTBN key.
+  WRITE_CTRL(OTBN, GENERATE_HW);
+
+  // Start the operation and wait for it to complete.
+  keymgr_start(diversification);
+  return keymgr_wait_until_done();
+}
+
+/**
+ * Clear the requested sideload slot.
+ *
+ * The `slot` parameter should be one of:
+ *   - KEYMGR_SIDELOAD_CLEAR_VAL_VALUE_AES
+ *   - KEYMGR_SIDELOAD_CLEAR_VAL_VALUE_KMAC
+ *   - KEYMGR_SIDELOAD_CLEAR_VAL_VALUE_OTBN
+ *
+ * @param slot Value to write to the SIDELOAD_CLEAR register.
+ */
+static status_t keymgr_sideload_clear(uint32_t slot) {
+  HARDENED_TRY(keymgr_is_idle());
+
+  // Ensure that the entropy complex has been initialized.
+  HARDENED_TRY(entropy_complex_check());
+
+  // Set SIDELOAD_CLEAR to begin continuously clearing the requested slot.
+  abs_mmio_write32(
+      kBaseAddr + KEYMGR_SIDELOAD_CLEAR_REG_OFFSET,
+      bitfield_field32_write(0, KEYMGR_SIDELOAD_CLEAR_VAL_FIELD, slot));
+
+  // Spin for 100 microseconds.
+  // TODO: this value seems to work for tests, but it would be good to run a
+  // more principled analysis.
+  busy_spin_micros(100);
```

Comment:
nit: For hardening, confirm that the expected value was written to `KEYMGR_SIDELOAD_CLEAR_REG_OFFSET`. I suspect that the action of reading back the register will give us enough time to perform the clear operation inside `keymgr`, but don't think it hurts to have the delay.